File: src/validate.py
#!/usr/bin/env python3
import json
import sys
import os

def validate():
    path = 'tasks.json'
    if not os.path.exists(path):
        print(f"Error: {path} not found.")
        sys.exit(1)

    try:
        with open(path, 'r') as f:
            data = json.load(f)
        
        errors = []
        
        for field in ['name', 'description', 'techStacks', 'roadmap']:
            if field not in data:
                errors.append(f"Missing required project field: '{field}'")
        
        if not errors:
            if not isinstance(data.get('techStacks'), list):
                errors.append("Field 'techStacks' must be a list.")
            if not isinstance(data.get('roadmap'), list):
                errors.append("Field 'roadmap' must be a list.")

        if not errors:
            roadmap = data['roadmap']
            for p_idx, phase in enumerate(roadmap):
                for p_field in ['id', 'description', 'tasks']:
                    if p_field not in phase:
                        errors.append(f"Phase {p_idx} missing field: '{p_field}'")
                
                if 'tasks' in phase:
                    tasks = phase['tasks']
                    if not isinstance(tasks, list):
                        errors.append(f"Phase {p_idx} 'tasks' must be a list.")
                    else:
                        for t_idx, task in enumerate(tasks):
                            for t_field in ['id', 'description', 'details', 'completed']:
                                if t_field not in task:
                                    errors.append(f"Task {t_idx} in Phase {p_idx} missing field: '{t_field}'")
                            
                            if 'id' in task and not isinstance(task['id'], int):
                                errors.append(f"Task {t_idx} in Phase {p_idx} 'id' must be an int.")
                            if 'description' in task and not isinstance(task['description'], str):
                                errors.append(f"Task {t_idx} in Phase {p_idx} 'description' must be a string.")
                            if 'details' in task and not isinstance(task['details'], list):
                                errors.append(f"Task {t_idx} in Phase {p_idx} 'details' must be a list.")
                            if 'completed' in task and not isinstance(task['completed'], bool):
                                errors.append(f"Task {t_idx} in Phase {p_idx} 'completed' must be a boolean.")

        if 'bugs' in data:
            bugs = data['bugs']
            if not isinstance(bugs, list):
                errors.append("Field 'bugs' must be a list.")
            else:
                for b_idx, bug in enumerate(bugs):
                    for b_field in ['id', 'description', 'details', 'suggestedSolution', 'fixed']:
                        if b_field not in bug:
                            errors.append(f"Bug {b_idx} missing field: '{b_field}'")
                    
                    if 'id' in bug and not isinstance(bug['id'], int):
                        errors.append(f"Bug {b_idx} 'id' must be an int.")
                    if 'description' in bug and not isinstance(bug['description'], str):
                        errors.append(f"Bug {b_idx} 'description' must be a string.")
                    if 'details' in bug and not isinstance(bug['details'], list):
                        errors.append(f"Bug {b_idx} 'details' must be a list.")
                    if 'suggestedSolution' in bug and not isinstance(bug['suggestedSolution'], str):
                        errors.append(f"Bug {b_idx} 'suggestedSolution' must be a string.")
                    if 'fixed' in bug and not isinstance(bug['fixed'], bool):
                        errors.append(f"Bug {b_idx} 'fixed' must be a boolean.")

        if errors:
            print("Schema Validation Failed:")
            for err in errors:
                print(f" - {err}")
            sys.exit(1)

        print("tasks.json verified successfully.")
        sys.exit(0)
    except json.JSONDecodeError as e:
        print(f"Failed to parse tasks.json: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"An unexpected error occurred during validation: {e}")
        sys.exit(1)

if __name__ == '__main__':
    validate()

File: src/validate.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

/**
 * Represents a successful or failed operation.
 */
export type Result<T, E = Error> =
  | { ok: true; value: T }
  | { ok: false; error: E };

/**
 * Domain error for validation failures.
 */
export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Type guard to check if an unknown value is a plain object.
 *
 * @param val - The value to check.
 * @returns True if the value is a non-null object, false otherwise.
 */
function isObject(val: unknown): val is Record<string, unknown> {
  return typeof val === 'object' && val !== null && !Array.isArray(val);
}

/**
 * Parses JSON content safely, returning a Result instead of throwing.
 *
 * @param content - The raw JSON string.
 * @returns A Result containing the parsed object or a ValidationError.
 */
function parseJson(content: string): Result<unknown, ValidationError> {
  try {
    return { ok: true, value: JSON.parse(content) };
  } catch (error) {
    return { ok: false, error: new ValidationError(`Failed to parse tasks.json: ${error}`) };
  }
}

/**
 * Validates the core required fields of the project root.
 *
 * @param payload - The parsed root JSON object.
 * @returns A Result indicating success or a validation error.
 */
function validateRootFields(payload: Record<string, unknown>): Result<void, ValidationError> {
  const requiredTopFields = ['name', 'description', 'techStacks', 'roadmap'];
  for (const field of requiredTopFields) {
    if (!(field in payload)) {
      return { ok: false, error: new ValidationError(`Missing required project field: '${field}'`) };
    }
  }

  if (!Array.isArray(payload.techStacks)) {
    return { ok: false, error: new ValidationError("Field 'techStacks' must be an array.") };
  }

  if (!Array.isArray(payload.roadmap)) {
    return { ok: false, error: new ValidationError("Field 'roadmap' must be an array.") };
  }

  return { ok: true, value: undefined };
}

/**
 * Validates an individual task within a phase.
 *
 * @param task - The task object to validate.
 * @param taskIndex - The index of the task.
 * @param phaseIndex - The index of the parent phase.
 * @returns A Result indicating success or a validation error.
 */
function validateTask(task: unknown, taskIndex: number, phaseIndex: number): Result<void, ValidationError> {
  if (!isObject(task)) {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} must be an object.`) };
  }

  const requiredFields = ['id', 'description', 'details', 'tests', 'completed'];
  for (const field of requiredFields) {
    if (!(field in task)) {
      return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} missing field: '${field}'`) };
    }
  }

  if (typeof task.id !== 'number') {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} 'id' must be a number.`) };
  }
  if (typeof task.description !== 'string') {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} 'description' must be a string.`) };
  }
  if (!Array.isArray(task.details)) {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} 'details' must be an array.`) };
  }
  if (!Array.isArray(task.tests)) {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} 'tests' must be an array.`) };
  }
  if (typeof task.completed !== 'boolean') {
    return { ok: false, error: new ValidationError(`Task ${taskIndex} in Phase ${phaseIndex} 'completed' must be a boolean.`) };
  }

  return { ok: true, value: undefined };
}

/**
 * Validates an individual phase within the roadmap.
 *
 * @param phase - The phase object to validate.
 * @param phaseIndex - The index of the phase.
 * @returns A Result indicating success or a validation error.
 */
function validatePhase(phase: unknown, phaseIndex: number): Result<void, ValidationError> {
  if (!isObject(phase)) {
    return { ok: false, error: new ValidationError(`Phase ${phaseIndex} must be an object.`) };
  }

  const requiredPhaseFields = ['id', 'description', 'tasks', 'tests'];
  for (const field of requiredPhaseFields) {
    if (!(field in phase)) {
      return { ok: false, error: new ValidationError(`Phase ${phaseIndex} missing field: '${field}'`) };
    }
  }

  if (!Array.isArray(phase.tasks)) {
    return { ok: false, error: new ValidationError(`Phase ${phaseIndex} 'tasks' must be an array.`) };
  }

  if (!Array.isArray(phase.tests)) {
    return { ok: false, error: new ValidationError(`Phase ${phaseIndex} 'tests' must be an array.`) };
  }

  for (let taskIndex = 0; taskIndex < phase.tasks.length; taskIndex++) {
    const taskResult = validateTask(phase.tasks[taskIndex], taskIndex, phaseIndex);
    if (!taskResult.ok) {
      return taskResult;
    }
  }

  return { ok: true, value: undefined };
}

/**
 * Validates the full roadmap array structure.
 *
 * @param roadmap - The array of phases.
 * @returns A Result indicating success or a validation error.
 */
function validateRoadmap(roadmap: unknown[]): Result<void, ValidationError> {
  for (let phaseIndex = 0; phaseIndex < roadmap.length; phaseIndex++) {
    const phaseResult = validatePhase(roadmap[phaseIndex], phaseIndex);
    if (!phaseResult.ok) {
      return phaseResult;
    }
  }
  return { ok: true, value: undefined };
}

/**
 * Validates the optional bugs array.
 *
 * @param bugs - The bugs array to validate.
 * @returns A Result indicating success or a validation error.
 */
function validateBugs(bugs: unknown): Result<void, ValidationError> {
  if (!Array.isArray(bugs)) {
    return { ok: false, error: new ValidationError("Field 'bugs' must be an array.") };
  }

  for (let bugIndex = 0; bugIndex < bugs.length; bugIndex++) {
    const bug = bugs[bugIndex];
    if (!isObject(bug)) {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} must be an object.`) };
    }

    const requiredFields = ['id', 'description', 'details', 'suggestedSolution', 'fixed'];
    for (const field of requiredFields) {
      if (!(field in bug)) {
        return { ok: false, error: new ValidationError(`Bug ${bugIndex} missing field: '${field}'`) };
      }
    }

    if (typeof bug.id !== 'number') {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} 'id' must be a number.`) };
    }
    if (typeof bug.description !== 'string') {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} 'description' must be a string.`) };
    }
    if (!Array.isArray(bug.details)) {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} 'details' must be an array.`) };
    }
    if (typeof bug.suggestedSolution !== 'string') {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} 'suggestedSolution' must be a string.`) };
    }
    if (typeof bug.fixed !== 'boolean') {
      return { ok: false, error: new ValidationError(`Bug ${bugIndex} 'fixed' must be a boolean.`) };
    }
  }

  return { ok: true, value: undefined };
}

/**
 * Main entry point to validate tasks.json file.
 *
 * @param filePath - The absolute path to tasks.json.
 * @returns A Result indicating if the entire file matches the schema.
 */
export function validateTasksJson(filePath: string): Result<void, ValidationError> {
  if (!fs.existsSync(filePath)) {
    return { ok: false, error: new ValidationError(`File ${filePath} not found.`) };
  }

  const rawContent = fs.readFileSync(filePath, 'utf8');
  const parsedResult = parseJson(rawContent);
  
  if (!parsedResult.ok) {
    return parsedResult;
  }

  const payload = parsedResult.value;
  if (!isObject(payload)) {
    return { ok: false, error: new ValidationError("Root of tasks.json must be an object.") };
  }

  const rootResult = validateRootFields(payload);
  if (!rootResult.ok) return rootResult;

  const roadmapResult = validateRoadmap(payload.roadmap as unknown[]);
  if (!roadmapResult.ok) return roadmapResult;

  if ('bugs' in payload) {
    const bugsResult = validateBugs(payload.bugs);
    if (!bugsResult.ok) return bugsResult;
  }

  return { ok: true, value: undefined };
}

// CLI Execution Wrapper
if (import.meta.url === `file://${process.argv[1]}`) {
  const currentFilePath = fileURLToPath(import.meta.url);
  const currentDirPath = path.dirname(currentFilePath);
  const tasksPath = path.resolve(currentDirPath, '../app/tasks.json');
  
  const result = validateTasksJson(tasksPath);
  if (!result.ok) {
    console.error("Schema Validation Failed:");
    console.error(` - ${result.error.message}`);
    process.exit(1);
  }
  
  console.log("tasks.json verified successfully.");
  process.exit(0);
}
